### ISSUE TITLE
Controller should retry kubectl apply during syncing

### ISSUE THREAD

**Issue by alexmt** [argoproj/argo-cd#1532]
**Describe the bug**
The kubectl apply might fail due to intermittent errors. Especially if it is trying to apply changes to CRD backed by third party api server. Right now application sync fails even if single `kubectl apply` fails. It was observed that application which includes multiple `calico/GlobalNetworkPolicy` resources never synced successfully because at least one resource fails during syncing.

As a solution, we can introduce retries during syncing. This is not a common issue and it is difficult to differentiate the retriable error from non-retriable. So by default reties should be disabled and can be enabled per-application using `spec.syncPolicy.retryCount` field. E.g.

```
apiVersion: argoproj.io/v1alpha1
kind: Application
metadata:
  name: guestbook
spec:
  syncPolicy:
    retryCount: 3
```

**Comment by alexmt** [comment]
After discussion with @jessesuen we've identified one more possibility. The error might happen because Argo CD is trying to run `kubectl apply`  in parallel and calico. In this case, retry might make the problem worse. Instead we might have to limit the number of parallel `kubectl apply` per group/kind

**Comment by alexec** [comment]
Leave this one with me. I'm doing work on the sync logic.

**Comment by alexec** [comment]
@alexmt do we want to pursue this anymore?

**Comment by jessesuen** [comment]
> @alexmt do we want to pursue this anymore?

I think now that we surface the exit code, it should be enough. Would prefer to mark this wontfix.

**Comment by jessesuen** [comment]
Closing this. As a separate issue, I do believe auto-sync needs retry functionality (which is different than self-healing).

**Comment by sevteen** [comment]
Is there any option to hint argocd to retry the failed (auto) syncs? As mentioned here in the original description, I have a case when CustomResources managed by 3rd party API can't be created initially, but manual sync after a couple of seconds succeeds. I've tried to enable Self-heal and that didn't help, but I guess self-heal isn't designed for such a scenario? (see [Automated sync semantics](https://argoproj.github.io/argo-cd/user-guide/auto_sync/#automated-sync-semantics))

> Automatic sync will not reattempt a sync if the previous sync attempt against the same commit-SHA and parameters had failed. 

<br/>

Just FYI my application consists of [ClusterIssuer](https://cert-manager.io/docs/concepts/issuer/) CRs together with [cert-manager](https://cert-manager.io/docs/installation/kubernetes/#installing-with-regular-manifests) and I get the following error during the sync (only initially):

	
	Internal error occurred: failed calling webhook "webhook.cert-manager.io": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.245.18.43:443: connect: connection refused

I can say that cert-manager is created first and then ClusterIssuer. I also tried to increase the window between cert-manager and CR creation, by assigning later to a `PostSync` phase (with non-zero [sync-wave](https://argoproj.github.io/argo-cd/user-guide/sync-waves/#how-do-i-configure-waves) value) but it didn't help.


This makes CI/CD automation impossible since I have to manually sync the failed app every time the sync is made against a fresh k8s test cluster (as part of the CI pipeline). I'm using v1.6.1.


P.S. a side question: sync-wave controls the order of resource hook execution within the app, order isn't guaranteed across the apps right? Is there any way to order the application syncs? Thanks!

**Comment by zioproto** [comment]
@tsobe I am facing your same issue. What is the best way ?  Create a dedicated application for the ClusterIssuer crd ? thanks

**Comment by tsobe** [comment]
> @tsobe I am facing your same issue. What is the best way ? Create a dedicated application for the ClusterIssuer crd ? thanks

Hey @zioproto! Yes as far as I remember. Feel free to check [cert-issuers ](https://github.com/tsobe/lobiani/tree/master/infra-config/apps/cert-issuers), [cert-manager](https://github.com/tsobe/lobiani/tree/master/infra-config/apps/cert-manager) and [workaround in the CI](https://github.com/tsobe/lobiani/blob/master/.circleci/config.yml#L280-L292) for more details.
